ROX-25354: Feature flags `README.md` update (#12226)
stackrox/stackrox: pkg/features/README.md

Feature flags can be valuable to ship features in a preview state, to provide the end user a way to disable some functionality or to control any boolean setting.

## Use cases

### Feature release stages

> To promote a feature through various the release stages, consider the following progression:

Reviewer: ```suggestion
To promote a feature through the release stages, consider the following progression:
```
Sorry I think I introduced this typo myself.

> To promote a feature through various release stages, consider the following progrssion of the feature flag changes:

Reviewer: ```suggestion
To promote a feature through various the release stages, consider the following progression:
```

1. Until the feature is ready to be tested by customers, the flag should be disabled by default and unchangeable in production.
2. The flag, disabled by default, can be made changeable to allow customers share feedback on the feature.
3. When the feature is complete and generally available, the flag can be enabled by default.

> 4. After the feature has been tested in production for some time, but not longer, remove the flag.

Reviewer 2: nit: wording "tested in production" feels contradictory for me. How about "active in production" instead.

Author: This is a punch to the idea of having a flag enabled by default. Thanks for noticing!

Author: FTR, the time the flag is enabled in production is the time when we look at how the feature is working before removing the flag with a raised hand over the flag deactivation button. Kinda testing.

### Feature deprecation

> When an existing feature needs to be removed, a feature flag could be used to gradually deprecate the feature:

Reviewer: Could or should?

Author: ```suggestion
When an existing feature needs to be removed, a feature flag should be used to gradually deprecate the feature:
```
Right, should suits better.

> 1. Create a feature flag, disabled by default, with a variable having `HIDDEN`, `SUPPRESSION` or similar word in the name. Do not use `DISABLED`.

Reviewer: I'd suggest to add the reason why "DISABLED" should be avoided. I suppose it's to avoid a situation like `FEATURE_DISABLED.Enabled()`?

> 1. Create a feature flag, disabled by default, with a variable having `HIDDEN`, `SUPPRESSION` or similar word in the name. Do not use `DISABLED`, because the condition for the enabled feature will read bad.

Reviewer: ```suggestion
1. Create a feature flag, disabled by default, with a variable having `HIDDEN`, `SUPPRESSION` or similar word in the name. Do not use `DISABLED`, because the condition for the enabled feature will read badly.
```

2. Put the functionality being deprecated under condition, e.g., `if !featureSuppression.Enabled() { feature }`, and announce the deprecation.
3. After some grace period, make the flag enabled by default and announce removal.

### Dependencies

This package does not implement any support for managing dependencies between flags. Make sure to document potential dependencies close to the flag definition.

> :warning: Feature flags cannot be used inside migrations or schema changes.
> Migrations must be merged to the master branch without any feature flag gate, and must not break any current features.

## Adding a feature flag

> To add a feature flag, add a variable with your feature to `list.go`. To register this feature flag variables, you are required to provide the following:

Reviewer: ```suggestion
To add a feature flag, add a variable with your feature to `list.go`. To register feature flag variables, you are required to provide the following:
```

* Name: This is a short description of what this flag is about.

> * Environment Variable name: This is the environment variable which needs to be set to override the default value of this flag. Env var names **must** start with `ROX_`.

Reviewer: ```suggestion
* Environment variable name: This is the environment variable which users set to override the default value of this flag. Env var names **must** start with `ROX_`.
```

* Options:
  * `unchangeableInProd` makes the flag unchangeable in release builds. On development builds, the setting _can_ be changed. Consider using this option in the early stage of feature development.

>   * `enabled` makes the flag enabled by default. Use this option when the feature is complete, meet GA requirements, and needs to be enabled.

Reviewer: ```suggestion
  * `enabled` makes the flag enabled by default. Use this option when the feature is complete, meets GA requirements, and needs to be enabled.
```
